{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ERaBLE\n",
    "\n",
    "## 0. Introduction\n",
    "\n",
    "The problem addressed by Binet _et al._ (2016) is central to phylogenomics and involves the estimation of species trees based on a large set of gene trees. However they do not address the topological aspect of this problem but focus on how to get meaningful branch length estimates for the species tree based on a set of homologous genomic features."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The problem is an equality constrained continuous optimization problem. The objective is a weighted least squares criterion (WLS) and the problem translates to a classic quadratic programming problem that can be solved using Lagrange multipliers. Least squares problems are known to be convex. The problem can be stated as:\n",
    "\n",
    "** Minimize: **\n",
    "$$\n",
    "    Q(\\hat{\\alpha},\\hat{b}) = \\sum\\limits_{k=1}^m\\sum\\limits_{\\{i,j\\} \\subset L_k} w_{ij}^{(k)}(\\hat{\\alpha}_k\\delta_{ij}-\\hat{d}_{ij})^2  \n",
    "$$\n",
    "\n",
    "** subject to: **\n",
    "$$\n",
    "    \\sum\\limits_{k=1}^mZ_k\\hat{\\alpha}_k = \\sum\\limits_{k=1}^{m}Z_k\n",
    "$$\n",
    "\n",
    "Where $L_k$ is the set of taxa for which sequences of $G_k$ (gene $k$) are available, $\\delta_{ij}^{(k)}$ the distance between taxon $i$ and $j$ for gene $k$ (from input distance matrix), $\\hat{b}$ the estimated branch length, $\\hat{\\alpha}_k$ the scale factor, with $\\hat{\\alpha}_k = 1/\\hat{r}_k$ with $\\hat{r}_k$ the evolutionary rate of $G_k$.\n",
    "\n",
    "This is the general form of the constraint, ERaBLE chooses $Z_k = N_k\\sum\\limits_{i,j \\in L_k}\\delta_{ij}^{(k)}$\n",
    "with $N_k$ the alignment length of $G_k$.\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The problem can however be rephrased in matrix notation (Additional File 1 of Binet _et al._ 2016). Let $\\hat{\\alpha} = (\\hat{\\alpha}_1, \\hat{\\alpha}_2, \\dots, \\hat{\\alpha}_m)^T$ and similarly $\\hat{b} = (\\hat{b}_1,\\hat{b}_2, \\dots, \\hat{b}_\\tau)^T$ with $\\tau = |E(\\mathcal{T})|$ where $\\mathcal{T}$ denotes the tree topology graph. Let $\\delta_k$ be the vector with all pairwise distances $\\delta_{ij}^{(k)}$ for $G_k$. Similarly, let $\\hat{d}$ be the vector with all additive distances from $\\hat{b}$ and let $\\hat{d}_k$ denote the vector of all additive distances for the taxa included in $L_k$.\n",
    "\n",
    "We now define the topological matrix $A$ that represents the topology $\\mathcal{T}$ as the $n(n-1)/2 \\times \\tau$ matrix where each row corresponds to a pair of taxa and each column to a branch in $\\mathcal{T}$. Let $a_{ij,e}$ be the element for pair of taxa $i$ and $j$ for branch $e$, $a_{ij,e}$ is set to $1$ if $e$ is on the path between $i$ and $j$, and to $0$ otherwise. Similar as with the definition of $\\hat{d}_k$, let $A_k$ be the $|L_k|(|L_k|-1)/2 \\times \\tau$ matrix obtained by removing all pairs of taxa that are not present in in $L_k$. Now $\\hat{d} = A\\hat{b}$ and $\\hat{d}_k = A_k\\hat{b}$.\n",
    "\n",
    "Let $W_k$, the weight matrix, be the square matrix with as diagonal elements $w_{ij}^{(k)}$ and all other elements set to $0$. Finally let $z = (Z_1,Z_2,\\dots,Z_m)^T$ and $Z=\\sum\\limits_{k=1}^{m}Z_k$.\n",
    "\n",
    "The WLS problem can now be expressed as:\n",
    "\n",
    "** Minimize: **\n",
    "$$\n",
    "    Q(\\hat{\\alpha},\\hat{b}) = \\sum\\limits_{k=1}^m(\\hat{\\alpha}_k\\delta_k - A_k\\hat{b})^TW_k(\\hat{\\alpha}_k\\delta_k - A_k\\hat{b})\n",
    "$$\n",
    "\n",
    "** Subject to: **\n",
    "$$\n",
    "    z^T\\hat{\\alpha} = Z\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This problem can be solved using the method of Lagrange multipliers. The Lagrangian function is:\n",
    "\n",
    "$$\\mathcal{L}(\\hat{\\alpha},\\hat{b},\\lambda) = Q(\\hat{\\alpha},\\hat{b})+\\lambda(z^T\\hat{\\alpha}-Z)$$\n",
    "\n",
    "A necessary and sufficient condition for the tuple $(\\hat{\\alpha}, \\hat{b})$ to be an optimal solution for the above problem is that the gradient of the Lagrangian is zero _i.e._ $\\nabla_{\\hat{\\alpha},\\hat{b},\\lambda}\\mathcal{L}(\\hat{\\alpha}, \\hat{b}, \\lambda) = 0$, note that this is sufficient since the WLS problem is convex. We get the system of equations:\n",
    "\n",
    "$$\n",
    "    \\frac{\\partial\\mathcal{L}}{\\partial\\hat{\\alpha}_k} = 0\n",
    "$$\n",
    "$$\n",
    "    \\frac{\\partial\\mathcal{L}}{\\partial\\hat{b}} = 0\n",
    "$$\n",
    "$$\n",
    "    \\frac{\\partial\\mathcal{L}}{\\partial\\lambda} = 0\n",
    "$$\n",
    "\n",
    "Which results in:\n",
    "\n",
    "$$\n",
    "    \\hat{\\alpha}_k\\delta_k^TW_k\\delta_k - \\delta^TW_kA_kb + \\lambda z^T/2 = 0\n",
    "$$\n",
    "$$\n",
    "    \\sum\\limits_{k=1}^{m}(A_k^TW_kA_k\\hat{b} - \\hat{\\alpha}_kA_k^tW_k\\delta_k) = 0\n",
    "$$\n",
    "$$\n",
    "    z^T\\hat{\\alpha} -Z = 0\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here the first equation is a system of $m$ equations (with $m$ the number of genes), the second a system of $\\tau$ equations (with $\\tau$ the number ofb branches) and the last consist of one equation. This can be rewritten as:\n",
    "\n",
    "$$\n",
    "    D\\hat{\\alpha} + B^T\\hat{b} + \\lambda z = 0\n",
    "$$\n",
    "$$\n",
    "    B\\hat{\\alpha} + C\\hat{b} = 0\n",
    "$$\n",
    "$$\n",
    "    z^T\\hat{\\alpha}=Z\n",
    "$$\n",
    "With $D = \\delta_k^TW_k\\delta_k$, $B^T = - \\delta^T_kW_kA_k$ and $C = \\sum\\limits_{k=1}^mA^T_kW_kA_k$. The $1/2$ coeffici&euml;nt was dropped, as we are not interested in the value of the Lagrange multiplier. This system can be solved naively by matrix multiplication in $O(mn^4)$ time. Note that $Z_k = N_k\\sum\\limits_{i,j \\in L_k}\\delta_{ij}^{(k)}$ with $N_k$ the alignment length of $G_k$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Calculation of distances"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some functions to get distance matrices from a bunch of multiple sequence alignments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from numpy import inf\n",
    "from heapq import heappop, heappush, heappushpop, heapify\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "import itertools\n",
    "from ete3 import Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "species = {\n",
    "    'arath': re.compile('AT[^R].+'),\n",
    "    'ambtr': re.compile('ATR.+'),\n",
    "    'chlre': re.compile('CR.+'),\n",
    "    'phypa': re.compile('PP.+'),\n",
    "    'poptr': re.compile('PT.+'),\n",
    "    'orysa': re.compile('OS.+'),\n",
    "    'zeama': re.compile('ZM.+'),\n",
    "    'zosma': re.compile('Zosma.+'),\n",
    "    'ulvmu': re.compile('evm.+'),\n",
    "    'vitvi': re.compile('VV.+'),\n",
    "    'thepa': re.compile('TP.+'),\n",
    "    'bradi': re.compile('BD.+'),\n",
    "    'spipo': re.compile('Spipo.+'),\n",
    "    'caros': re.compile('Caros.+'),\n",
    "    'ostta': re.compile('OT.+')\n",
    "}\n",
    "\n",
    "def proportion_different_sites(s1, s2):\n",
    "    \"\"\"\n",
    "    Calculate the proportion of different sites between pairwise sequences\n",
    "    :param s1:\n",
    "    :param s2:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    p = 0\n",
    "    for i in range(len(s1)):\n",
    "        if s1[i] != s2[i]:\n",
    "            p += 1\n",
    "    return p/len(s1)\n",
    "\n",
    "\n",
    "def jukes_cantor(s1, s2):\n",
    "    \"\"\"\n",
    "    Implementation of the JC69 model (Jukes & Cantor, 1969) which assumes\n",
    "    that every nucleotide has the same instantaneous rate of changing into another\n",
    "    nucleotide (an assumption that is almost never valid).\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    p = proportion_different_sites(s1, s2)\n",
    "    d = -(3/4)*np.log(1-(4/3)*p)\n",
    "    var_d = (p*(1-p)/len(s1))*(1/(1-4*p/3))\n",
    "    return d, var_d\n",
    "\n",
    "\n",
    "def distance_matrix(msa, distance='JC69'):\n",
    "    \"\"\"\n",
    "    Function to get a distance matrix from a multiple sequence alignment\n",
    "    :param msa: dictionary with aligned sequences\n",
    "    :param distance: distance metric to use (currently only JC69 supported)\n",
    "    :return: distance matrix (pandas data frame)\n",
    "    \"\"\"\n",
    "    l = list(msa.keys())\n",
    "    df = pd.DataFrame(np.zeros((len(l),len(l))), index=l, columns=l)\n",
    "    \n",
    "    for i in range(len(list(df.index))):\n",
    "        gene_1 = df.index[i]\n",
    "        \n",
    "        for j in range(i+1,len(list(df.index))):\n",
    "            gene_2 = df.index[j]\n",
    "            \n",
    "            if distance == 'JC69':\n",
    "                d = jukes_cantor(msa[gene_1], msa[gene_2])\n",
    "                df[gene_1][gene_2] = d[0]\n",
    "                df[gene_2][gene_1] = d[0]\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "def read_msa(msa):\n",
    "    \"\"\"\n",
    "    Read multiple sequence alignment (MSA) in PHYLIP format.\n",
    "    :param msa: MSA file\n",
    "    :return: dictionary\n",
    "    \"\"\"\n",
    "    msa_dict = {}\n",
    "    \n",
    "    with open(msa, 'r') as f:\n",
    "        content = f.readlines()\n",
    "        \n",
    "    n_seq, length = int(content[0].split()[0].strip()), int(content[0].split()[1].strip())\n",
    "\n",
    "    for i in range(1,n_seq*2,2):\n",
    "        msa_dict[content[i].strip()] = content[i+1].strip()\n",
    "            \n",
    "    return msa_dict\n",
    "\n",
    "\n",
    "def get_species(gene, species):\n",
    "    \"\"\"\n",
    "    Get species for a gene using regex matches.\n",
    "    \"\"\"\n",
    "    for sp, p in species.items():\n",
    "        if p.match(gene):\n",
    "            return sp\n",
    "    raise ValueError(\"species not found for gene {}!\".format(gene))\n",
    "    \n",
    "\n",
    "def collapse_on_species(distance_matrix, species):\n",
    "    \"\"\"\n",
    "    Collapse a given distance matrix for a gene family on species level.\n",
    "    Distances for duplicates within a species are averaged.\n",
    "    \"\"\"\n",
    "    matrix = pd.DataFrame()\n",
    "    sp_set = set()\n",
    "    to_drop = []\n",
    "    for gene in distance_matrix.index:\n",
    "        sp = get_species(gene, species)\n",
    "        if sp in sp_set:\n",
    "            matrix[sp] += distance_matrix[gene]\n",
    "            matrix[sp] /= 2\n",
    "            to_drop.append(gene)\n",
    "        else:\n",
    "            matrix[sp] = distance_matrix[gene]\n",
    "            sp_set.add(sp)\n",
    "        \n",
    "    matrix = matrix.drop(to_drop)\n",
    "    matrix.index = matrix.columns\n",
    "    \n",
    "    for k in range(len(matrix.index)):\n",
    "        matrix.iloc[k,k] = 0\n",
    "    return matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Helper functions for ERaBLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def topology_matrix_dijkstra(tree, species):\n",
    "    \"\"\"\n",
    "    Input: newick tree\n",
    "    Output: topology matrix\n",
    "    \"\"\"\n",
    "    t = Tree(tree)\n",
    "    i = len(species)\n",
    "    for node in t.traverse():\n",
    "        if node.name in species:\n",
    "            node.name = species[node.name]\n",
    "        else:\n",
    "            node.name = i\n",
    "            i+=1\n",
    "    \n",
    "    graph = tree_to_adj_list(t)\n",
    "    leaves = list(t.get_leaves())\n",
    "    paths = {}\n",
    "    branches = set()\n",
    "    for i in range(len(leaves)):\n",
    "        node1 = leaves[i]\n",
    "        for j in range(i+1, len(leaves)):\n",
    "            node2 = leaves[j]\n",
    "            path, distance = dijkstra(graph, node1.name, node2.name)\n",
    "            path = [(path[x], path[x+1]) for x in range(len(path)-1)]\n",
    "            paths[tuple(sorted([node1.name, node2.name]))] = path\n",
    "            for p in path:\n",
    "                if p not in branches:\n",
    "                    branches.add(p)\n",
    "            \n",
    "    # put in a matrix\n",
    "    df = pd.DataFrame(index=list(paths.keys()), columns=list(branches))\n",
    "    for tup, path in paths.items():\n",
    "        for branch in path:\n",
    "            df[branch][tup] = 1\n",
    "    df = df.fillna(0)\n",
    "    df.sort_index(0, inplace=True)\n",
    "    df.sort_index(1, inplace=True)\n",
    "    return df, t\n",
    "\n",
    "\n",
    "def dijkstra(graph, source, sink=None):\n",
    "    \"\"\"\n",
    "    Implementation of Dijkstra's shortest path algorithm\n",
    "    Inputs:\n",
    "        - graph : dict representing the weighted graph\n",
    "        - source : the source node\n",
    "        - sink : the sink node (optional)\n",
    "    Ouput:\n",
    "        - distance : dict with the distances of the nodes to the source\n",
    "        - came_from : dict with for each node the came_from node in the shortest\n",
    "                    path from the source\n",
    "    \"\"\"\n",
    "    distance = {v: inf for v in graph}\n",
    "    distance[source] = 0\n",
    "    current = source\n",
    "    previous = {}\n",
    "    Q = [(source, 0)]\n",
    "    \n",
    "    while Q:\n",
    "        U = heappop(Q)\n",
    "        if sink and U == sink:\n",
    "            break\n",
    "        for V in graph[U[0]]:\n",
    "            alt = distance[U[0]] + V[0]\n",
    "            if alt < distance[V[1]]:\n",
    "                distance[V[1]] = alt\n",
    "                previous[V[1]] = U[0]\n",
    "                heappush(Q, (V[1], alt))\n",
    "                \n",
    "    if sink is None:\n",
    "        return distance, previous\n",
    "    else:\n",
    "        return reconstruct_path(previous, source, sink), distance[sink]\n",
    "\n",
    "    \n",
    "def reconstruct_path(previous, source, sink):\n",
    "    \"\"\"\n",
    "    Reconstruct the path from the output of the Dijkstra algorithm\n",
    "    Inputs:\n",
    "            - previous : a dict with the came_from node in the path\n",
    "            - source : the source node\n",
    "            - sink : the sink node\n",
    "    Ouput:\n",
    "            - the shortest path from source to sink (list)\n",
    "    \"\"\"\n",
    "    if sink not in previous:\n",
    "        return []\n",
    "    \n",
    "    V = sink\n",
    "    path = [V]\n",
    "    while V != source:\n",
    "        V = previous[V]\n",
    "        path.append(V)\n",
    "    return path\n",
    "\n",
    "\n",
    "def tree_to_adj_list(tree):\n",
    "    \"\"\"\n",
    "    Convert ete3 Tree object to an adjacency list representing the tree graph\n",
    "    \"\"\"\n",
    "    adj_list = {}\n",
    "    for node in tree.traverse('postorder'):\n",
    "        l = []\n",
    "        if not node.is_leaf():\n",
    "            # add children \n",
    "            l += [(1, c.name) for c in node.children]\n",
    "        if not node.is_root():\n",
    "            # add parent\n",
    "            l.append((1, node.up.name))\n",
    "        adj_list[node.name] = set(l)\n",
    "    return adj_list\n",
    "\n",
    "\n",
    "def code_species(species):\n",
    "    \"\"\"\n",
    "    Code species to integers\n",
    "    \"\"\"\n",
    "    if type(species) == dict:\n",
    "        species = list(species.keys())\n",
    "    species_to_int = {x: 0 for x in species}\n",
    "    i = 0\n",
    "    for x in sorted(species):\n",
    "        species_to_int[x] = i\n",
    "        i += 1\n",
    "    return species_to_int\n",
    "\n",
    "\n",
    "def sub_topology_matrix(subset, matrix):\n",
    "    \"\"\"\n",
    "    Get topology matrix for a subset of taxa\n",
    "    \"\"\"\n",
    "    to_drop = []\n",
    "    for pair in matrix.index:\n",
    "        if pair[0] not in subset or pair[1] not in subset:\n",
    "            to_drop.append(pair)\n",
    "    matrix = matrix.drop(to_drop)\n",
    "    return matrix\n",
    "\n",
    "\n",
    "def sequence_length(msa_file):\n",
    "    \"\"\"\n",
    "    Construct a weight matrix. Using alignment length as weight.\n",
    "    \"\"\"\n",
    "    with open(msa_file, 'r') as f:\n",
    "        content = f.readlines()\n",
    "        \n",
    "    length = int(content[0].split()[1].strip())\n",
    "    return length\n",
    "\n",
    "\n",
    "def weight_matrix(length, s):\n",
    "    \"\"\"\n",
    "    Construct a weight matrix. Using alignment length as weight.\n",
    "    \"\"\"\n",
    "    return np.diag([length for i in range(s)])\n",
    "\n",
    "\n",
    "def vectorize_delta(delta):\n",
    "    \"\"\"\n",
    "    Convert distance matrix to vector\n",
    "    \"\"\"\n",
    "    d = np.array(delta)\n",
    "    l = []\n",
    "    for i in range(0,len(d)):\n",
    "        for j in range(i+1, len(d)):\n",
    "            l.append(d[i,j]) \n",
    "    return np.array(l)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. ERaBLE implementation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def erable(tree_file, species, msa_files=None, distance_matrices=None, lengths=None):\n",
    "    if not msa_files and not distance_matrices:\n",
    "        raise ValueError(\"Provide either distance matrices or MSA files\")\n",
    "        \n",
    "    species_mapping = code_species(species)\n",
    "    \n",
    "    A, t = topology_matrix_dijkstra(tree_file, species_mapping)\n",
    "    \n",
    "    # construct C\n",
    "    C = np.zeros((A.shape[1], A.shape[1]))\n",
    "    weights = []\n",
    "    Aks = []\n",
    "    z = []\n",
    "    msas = []\n",
    "    Nks = []\n",
    "    delta_mats = []\n",
    "\n",
    "    if not msa_files:\n",
    "        msa_files = distance_matrices\n",
    "        \n",
    "    for k in range(len(msa_files)):\n",
    "        if not distance_matrices:\n",
    "            msa = read_msa(msa_files[k])\n",
    "            delta_matrix = collapse_on_species(distance_matrix(msa), species)\n",
    "        else:\n",
    "            delta_matrix = distance_matrices[k]\n",
    "        delta_matrix.index = [species_mapping[x] for x in delta_matrix.index]\n",
    "        delta_matrix.columns = list(delta_matrix.index)\n",
    "        delta_matrix.sort_index(0, inplace=True)\n",
    "        delta_matrix.sort_index(1, inplace=True)\n",
    "\n",
    "        if delta_matrix is not None and len(delta_matrix) > 1:\n",
    "            delta_mats.append(delta_matrix)\n",
    "            msas.append(msa_files[k])\n",
    "        \n",
    "    D = np.zeros((len(delta_mats), len(delta_mats)))\n",
    "    B = np.zeros((A.shape[1], len(delta_mats)))\n",
    "        \n",
    "    for k in range(len(delta_mats)):\n",
    "        delta = vectorize_delta(delta_mats[k]).reshape(-1,1)\n",
    "        Ak = sub_topology_matrix(set(delta_mats[k].index), A)\n",
    "        if lengths:\n",
    "            Nk = lengths[k]\n",
    "            Nks.append(Nk)\n",
    "        else:\n",
    "            Nk = sequence_length(msas[k])\n",
    "            Nks.append(Nk)\n",
    "        Wk = weight_matrix(Nk,len(delta))\n",
    "        C += Ak.T @ Wk @ Ak\n",
    "        weights.append(Wk)\n",
    "        Aks.append(Ak)\n",
    "        D[k,k] = delta.T @ Wk @ delta\n",
    "        B[:,k] = list(-Ak.T @ Wk @ delta)\n",
    "        Zk = Nk * delta.sum()\n",
    "        z.append(Zk)\n",
    "    \n",
    "    Z = sum(z)\n",
    "    \n",
    "    return D, B, z, C, Z\n",
    "    \n",
    "    # invert D, diagonal matrix so inverse is just diagonal elements^-1\n",
    "    D_inv = 1/D\n",
    "    D_inv[np.isinf(D_inv)] = 0\n",
    "    \n",
    "    z = np.array(z).reshape((-1,1))\n",
    "    u = B @ D_inv @ z\n",
    "    w = (z.reshape((1,-1)) @ D_inv @ z)[0][0]\n",
    "    M = C + (1/w) * u @ u.T - B @ D_inv @ B.T\n",
    "    \n",
    "    #return M, Z, w, u, z, B, D_inv, A, Nks\n",
    "    #return M, Z, w, u\n",
    "    b = np.linalg.solve(M, -(Z/w)*u)\n",
    "    alpha = D_inv @ ((z.reshape((-1,1)) @ u.reshape((1,-1)))/w - B.T) @ b + (Z/w)*D_inv @ z\n",
    "    #alpha = D_inv @ ((z.reshape((1,-1)) @ u.reshape((-1,1)))/w - B.T) @ b + (Z/w)*D_inv @ z\n",
    "    #alpha[alpha == 0] = 0.0001 # this shouldn't be necessary\n",
    "    c = (1/sum(Nks)) * sum([Nks[i]/alpha[i] for i in range(len(Nks))])\n",
    "    \n",
    "    r = 1/(c*alpha)\n",
    "    b = c * b\n",
    "\n",
    "    return(b, r, c, t)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "msa_files = [os.path.join('subset/', x) for x in sorted(os.listdir('subset/'))]\n",
    "b, r, c, tree = erable('tree.nwck', species, msa_files=msa_files)\n",
    "#M, Z, w, u, z, B, D_inv, A, Nks = erable(msa_files, 'tree.nwck', species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[  3.28167584e+01],\n",
       "       [  1.10677638e+01],\n",
       "       [  3.92583444e+00],\n",
       "       [  8.22513533e+00],\n",
       "       [  1.34470477e+03],\n",
       "       [  3.91750038e+00],\n",
       "       [  3.22303845e+02],\n",
       "       [  5.25676344e+00],\n",
       "       [  1.03627275e+01],\n",
       "       [  1.10656255e+01],\n",
       "       [  1.34466238e+03],\n",
       "       [  3.77135929e+00],\n",
       "       [  1.43907678e+00],\n",
       "       [  1.03796727e+01],\n",
       "       [  9.15167735e+02],\n",
       "       [ -1.45907169e+02],\n",
       "       [ -1.13572470e+02],\n",
       "       [ -3.21474737e+02],\n",
       "       [ -1.09077822e+03],\n",
       "       [ -3.23445998e+01],\n",
       "       [  1.13737427e+02],\n",
       "       [ -2.37942439e+01],\n",
       "       [ -1.34423909e+03],\n",
       "       [ -1.02246936e+03],\n",
       "       [  2.36863848e+01],\n",
       "       [  8.20169979e+00],\n",
       "       [ -5.61197928e-01],\n",
       "       [ -8.13629047e+00],\n",
       "       [ -6.94833054e+00],\n",
       "       [ -1.55796848e+01],\n",
       "       [ -7.82233675e+00],\n",
       "       [  9.13570774e-01],\n",
       "       [ -4.51979344e+00],\n",
       "       [ -9.88476068e+00],\n",
       "       [ -9.86404353e+00],\n",
       "       [  6.93193704e+00],\n",
       "       [ -1.21898092e+00],\n",
       "       [  1.58799031e+01],\n",
       "       [  2.50108193e+00],\n",
       "       [ -3.45701357e+00],\n",
       "       [  4.41250783e+00],\n",
       "       [  1.47497566e+00],\n",
       "       [ -3.72423508e+00],\n",
       "       [ -3.73292658e+00],\n",
       "       [ -2.48405289e+00],\n",
       "       [ -4.91163287e+00],\n",
       "       [ -1.46640957e+00],\n",
       "       [  6.06780213e+00],\n",
       "       [ -1.09329402e+01],\n",
       "       [ -1.09447794e+01],\n",
       "       [ -5.75782111e+00]])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Example from ERaBLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mats = []\n",
    "for f in os.listdir('../../erable1.0_Unix_Linux/Example/mats'):\n",
    "    m = pd.read_csv(os.path.join('../../erable1.0_Unix_Linux/Example/mats', f), sep=\"\\s+\", index_col=0, header=None)\n",
    "    m.columns = m.index\n",
    "    mats.append(m)\n",
    "\n",
    "lengths = [382, 210, 590]\n",
    "tree_file = '../../erable1.0_Unix_Linux/Example/input_tree_topology.nwk'\n",
    "tree_s = '((Tax10:0.51144,(Tax34:0.19411,(Tax20:0.19064,Tax4:0.15564)1:0.09995)1:0.09856),(Tax14:0.10233,Tax3:0.03293)1:0.01914);'\n",
    "species = [n.name for n in Tree(tree_s).get_leaves()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#erable(tree_file=tree_s, species=species, distance_matrices=mats, lengths=lengths)\n",
    "D, B, z, C, Z = erable(tree_file=tree_s, species=species, distance_matrices=mats, lengths=lengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def solve_constrained_quadratic_problem(P, q, A, b):\n",
    "    \"\"\"\n",
    "    Solve a linear constrained quadratic convex problem.\n",
    "    Inputs:\n",
    "        - P, q: quadratic and linear parameters of\n",
    "                the linear function to be minimized\n",
    "        - A, b: system of the linear constaints\n",
    "    Outputs:\n",
    "        - xstar: the exact minimizer\n",
    "        - vstar: the optimal Lagrance multipliers\n",
    "    \"\"\"\n",
    "    #p, n = A.shape  # size of the problem\n",
    "    n = A.shape[1]\n",
    "    A = np.bmat([[P,A.T],[A,np.zeros((A.shape[0],A.shape[0]))]])\n",
    "    B = np.bmat([[-q],[b]])\n",
    "    solution = np.linalg.solve(A,B)\n",
    "    xstar = solution[:n]\n",
    "    vstar = solution[n:]\n",
    "    return np.array(xstar), np.array(vstar)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "b = np.array([0.5]*C.shape[0])\n",
    "z = np.array(z).reshape(-1,1)\n",
    "alpha, l = solve_constrained_quadratic_problem(\n",
    "    D, (-B.T @ b).reshape(-1,1), z.T, np.array(Z).reshape(1,1)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[ 2030.378323],\n",
       "        [  676.73865 ],\n",
       "        [ 6059.52066 ],\n",
       "        [ 4028.844772]])"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solve_constrained_quadratic_problem()"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
